libyul/backends/evm/EVMCodeTransform.cpp

```diff
@@ -628,13 +655,72 @@ void CodeTransform::visitExpression(Expression const& _expression)
 	expectDeposit(1, height);
 }
 
+void CodeTransform::setupReturnVariablesAndFunctionExit()
+{
+	yulAssert(!m_functionExitStackHeight.has_value(), "");
+	yulAssert(m_scope, "");
+
+	ScopeGuard scopeGuard([oldScope = m_scope, this] { m_scope = oldScope; });
+	if (!m_scope->functionScope)
+	{
+		yulAssert(m_scope->superScope && m_scope->superScope->functionScope, "");
+		m_scope = m_scope->superScope;
+	}
+
+	if (m_delayedReturnVariables.empty())
+	{
+		m_functionExitStackHeight = m_evm15 ? 0 : 1;
+		return;
+	}
+
+	// Allocate slots for return variables as if they were declared as variables in the virtual function scope.
+	for (TypedName const& var: m_delayedReturnVariables)
+		(*this)(VariableDeclaration{var.location, {var}, {}});
```

Comment:
Note that this will add the correct source location to the return variable initialization, which I'd say is a nice touch.